modules/codegen/src/smithy4s/codegen/internals/SmithyBuild.scala

```diff
@@ -17,25 +17,87 @@
 package smithy4s.codegen
 package internals
 
-import io.circe.Codec
-import io.circe.generic.semiauto._
+import cats.syntax.all._
+import io.circe._
 import io.circe.syntax._
+import io.circe.generic.semiauto._
+import software.amazon.smithy.model.node.Node
+import software.amazon.smithy.openapi.OpenApiConfig
+
+import scala.collection.Seq
+import scala.reflect.ClassTag
+import scala.util.Try
 
 private[internals] final case class SmithyBuild(
     version: String,
-    imports: Seq[String],
-    maven: SmithyBuildMaven
-)
+    imports: Seq[os.FilePath],
+    plugins: Seq[SmithyBuildPlugin],
+    maven: Option[SmithyBuildMaven]
+) {
+  def getPlugin[T <: SmithyBuildPlugin](implicit
+      classTag: ClassTag[T]
+  ): Option[T] =
+    plugins.collectFirst { case t: T => t }
+}
+
 private[codegen] object SmithyBuild {
-  implicit val codecs: Codec[SmithyBuild] = deriveCodec
-  def writeJson(sb: SmithyBuild): String = sb.asJson.spaces4
+  // automatically map absence of value to empty Seq for ease of use
+  implicit def optionalSeqDecoder[T](implicit
+      base: Decoder[T]
+  ): Decoder[Seq[T]] =
+    Decoder.decodeOption(Decoder.decodeSeq[T]).map(_.getOrElse(Seq.empty))
+
+  implicit val pathDecoder: Decoder[os.FilePath] =
+    Decoder.decodeString.emapTry { raw =>
+      Try(os.FilePath(raw))
+    }
+
+  implicit val pluginDecoder: Decoder[Seq[SmithyBuildPlugin]] = Decoder
+    .decodeOption { (c: HCursor) =>
+      c.keys match {
+        case None => DecodingFailure("Expected JSON object", c.history).asLeft
+        case Some(keys) =>
+          keys.toList.traverse(key => c.get(key)(SmithyBuildPlugin.decode(key)))
+      }
+    }
+    .map(_.getOrElse(Seq.empty))
+
+  /* Class containing only the subset of the smithy-build.json properties that need
+   * to be serialized when creating a smithy-build.json file. Allows us to skip
+   * things that are both unnecessary and very complicated to serialize,
+   * such as OpenApiConfig.
+   */
```

Comment:
See comment - as discussed in the Discord, Smithy doesn't seem to think we'd ever need to serialize `OpenApiConfig` objects back to JSON. To be entirely fair to them, we _don't_ - the `writeJson` logic is about creating a `smithy-build.json` file from our SBT/Mill settings, and the openApi configuration isn't one of them. Still, it means that we can't use the same model for read and write.